module/os/linux/zfs/zfs_ioctl_os.c

```diff
@@ -280,14 +280,23 @@ openzfs_init(void)
 {
 	int err;
 	if ((err = zcommon_init()) != 0)
-		return (err);
+		goto zcommon_failed;
 	if ((err = icp_init()) != 0)
-		return (err);
+		goto icp_failed;
 	if ((err = zstd_init()) != 0)
-		return (err);
+		goto zstd_failed;
 	if ((err = openzfs_init_os()) != 0)
-		return (err);
+		goto openzfs_os_failed;
 	return (0);
+
+openzfs_os_failed:
+	zstd_fini();
+zstd_failed:
+	icp_fini();
+icp_failed:
+	zcommon_fini();
+zcommon_failed:
+	return (err);
```

Comment:
It seems to that after this refactoring we should be quite close to being able to initialize everything the same way on Linux and FreeBSD.  That said, it seems perfectly reasonable to keep this PR focused on the Linux side for now.